rasa/utils/tensorflow/models.py

```diff
@@ -47,6 +48,25 @@
 logger = logging.getLogger(__name__)
 
 
+def _merge_batch_outputs(
+    all_outputs: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+    batch_output: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+) -> Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]]:
+    if not all_outputs:
+        return batch_output
+    for key, val in batch_output.items():
+        if isinstance(val, np.ndarray):
+            all_outputs[key] = np.concatenate(
+                [all_outputs[key], batch_output[key]], axis=0
+            )
+
+        elif isinstance(val, dict):
+            # recurse and merge the inner dict first
+            all_outputs[key] = _merge_batch_outputs(all_outputs[key], val)
+
+    return all_outputs
```

Comment:
You can't really create it as a static method as well because it is recursive and hence needs `self` argument.

Reply:
Also, keeping it inside `RasaModel` felt unusual because it's more of a utility method which can be tested alone as well. Keeping it inside `RasaModel` means that we always invoke it as `rasa_model._merge_batch_outputs`.
What do you think?

Reply:
I think `RasaModel` currently contains a lot of these utility methods -- because they're specifically used by the class. After all, the method is tailored specifically for that class and can't be re-used in some other implementation of a model. So, including this method in `RasaModel` wouldn't break a pattern. And I think there's nothing wrong about calling it as `self._merge_batch_outputs` (from within `RasaModel` even when it's static) and as `RasaModel._merge_batch_outputs` from outside (when actually needed, currently only in tests).

Reply:
> After all, the method is tailored specifically for that class and can't be re-used in some other implementation of a model.

👍  I'll add it in.

> And I think there's nothing wrong about calling it as self._merge_batch_outputs (from within RasaModel even when it's static)

There is no `self` variable if the method is static.

Reply:
Okay, I changed the recursive call to `RasaModel._merge_batch_outputs`.

Reply:
> There is no self variable if the method is static.

Ah, I get you now. I meant when calling the method from within the model, not from within itself. You're right.